### ISSUE TITLE
Shrink ots receipts

### ISSUE THREAD

**Issue by LeoComandini** [opentimestamps/opentimestamps-client#88]
Once a timestamp is complete, a user may find useful to drop the redundant data included in the ots receipt, storing only the linear proof from the file hash to the *best* attestation. 

This feature has a draft implementation in [java-opentimestamps](https://github.com/opentimestamps/java-opentimestamps/blob/master/src/main/java/com/eternitywall/ots/Timestamp.java#L196) and it would be nice if also Python clients have access to it. However, if shrinking becomes a default procedure by clients, then new problems and/or risks may occur.

In case of a complete receipt with 4 branches (one for each calendar): 2 branches end in 2 `BitcoinBlockHeaderAttestation`s while the other 2 end in 2 `PendingAttestation`s. The shrunk receipt would consist in only the linear branch to the *best* `BitcoinBlockHeaderAttestation` (minimum height, in case of parity minimum length of the timestamp, in case of parity first one read). All the other three branches (along with the `PendingAttestation` in the remaining branch) are dropped.

It may occur that one of the two calendars whose `PendingAttestation` was not resolved actually has a better proof for the user. By dropping that branch, the user won't never be able to obtain that proof.

Furthermore, it may happen that one calendar lied and the `BitcoinBlockHeaderAttestation` doesn't actually match the merkleroot published in the corresponding block header. If that attestation is the *best* attestation, then the shrunk receipt would be completely useless.

Considering those issues, I was wondering which should be the correct way to set up this feature: 
- as a new command for the OpenTimestamps client, or
- as an independent Python script (stored somewhere else)

In the latter case, it would be something like:
```
python3 ots-shrink.py my-file.ots my-file.ots.shrunk [--notary {btc,ltc,eth}]
```
To avoid the problems mentioned above, the user should then check that the shrunk receipt `my-file.ots.shrunk` is actually valid before overwriting `my-file.ots`. A proof may contain attestations from different notaries, a shrunk receipt would have only the *best* attestation from the selected notary (default is `btc`). 

Summing up, does it make sense to implement this feature as a new command for the ots client, or it is too risky?
Moreover, are there other risks that I haven't pointed out?

**Comment by petertodd** [comment]
I think it makes sense as a feature within the ots client itself.

You're quite correct in that there is a potential risk in a calendar server lying, but that risk is easily mitigated by verifying the attestations yourself, against a trusted Bitcoin node, prior to deleting any data.

A possible UI could be something like the following:

    ots prune [--verify NOTARYSPEC] [--discard NOTARYSPEC] TIMESTAMP

With the args being:

- `-v, --verify NOTARYSPEC` Verify attestations from this notary prior to deleting anything; if the verification fails, do nothing. May be used multiple times. Default: `btc`
- `-d, --discard NOTARYSPEC` Discard attestations from this notary. Default: `pending:*`

A timestamp without any attestations is invalid, so the command will fail if there are no attestations left.

**Comment by LeoComandini** [comment]
Attestations are discarded in two cases:
- if the client specifies to discard certain attestations, with `--discard`
- if they are suboptimal (only *comparable* attestations, like `btc` or `ltc`, could be suboptimal)

With the design suggested, the client cannot specify whether or not to discard suboptimal attestations for a given notary. Should the client be able to do so?

For instance, suppose we have:
```
<digest>
 -> {ops}
    verify BitcoinBlockHeaderAttestation(1)
 -> {ops}
    verify BitcoinBlockHeaderAttestation(2)
 -> {ops}
    verify PendingAttestation(<calendar1>)
 -> {ops}
    verify PendingAttestation(<calendar2>)
```
If suboptimal attestations are always discarded, then there is no way to obtain:
```
<digest>
 -> {ops}
    verify BitcoinBlockHeaderAttestation(1)
 -> {ops}
    verify BitcoinBlockHeaderAttestation(2)
```
Is it fine, or should this case be managed?

**Comment by petertodd** [comment]
Re: suboptimal attestations, I think we can always combine them together when we're shrinking/pruning timestamps so long as we've verified them first.

**Comment by jonathancross** [comment]
Can this be closed now that we can prune timestamps?
